engine/user_interface/public/ui_element.hpp

```diff
@@ -0,0 +1,107 @@
+#pragma once
+
+#include "glm/vec2.hpp"
+#include "quad_draw_info.hpp"
+#include "spdlog/spdlog.h"
+
+#include <common.hpp>
+#include <memory>
+#include <stdint.h>
+
+class InputManager;
+/**
+ * Base class from which all ui elements inherit. Updating and submitting of the ui happens
+ * mostly in a hierarchical manner. each element calls its children's update and draw functions.
+ * class contains pure virtual functions.
+ */
+class UIElement
+{
+public:
+    explicit UIElement(uint16_t maxChildren)
+        : _maxChildren(maxChildren)
+    {
+    }
+
+    enum class AnchorPoint
+    {
+        eMiddle,
+        eTopLeft,
+        eTopRight,
+        eBottomLeft,
+        eBottomRight,
+    };
+
+    /**
+     * Whenever this gets called the updateChildrenAbsoluteLocations of the parent needs to get called as well!
+     * @param location the new location relative to the set anchor point of the parent.
+     */
+    void SetLocation(const glm::vec2& location) noexcept { _relativeLocation = location; }
+
+    /**
+     * @return the location of the element relative to the set anchorpoint of the parent element.
+     */
+    NO_DISCARD const glm::vec2& GetRelativeLocation() const noexcept { return _relativeLocation; }
+    NO_DISCARD const glm::vec2& GetAbsouluteLocation() const noexcept { return _absoluteLocation; }
+
+    NO_DISCARD const glm::vec2& GetScale() const noexcept { return _scale; }
+    void SetScale(const glm::vec2& scale) noexcept { _scale = scale; }
+
+    virtual void SubmitDrawInfo(MAYBE_UNUSED std::vector<QuadDrawInfo>& drawList) const = 0;
+
+    virtual void Update(const InputManager& input)
+    {
+        for (auto& i : _children)
+            i->Update(input);
+    }
+
+    UIElement* AddChild(std::unique_ptr<UIElement> child)
```

Comment:
Sharing the backing pointers of unique ptrs, sounds contradicting, shouldnt we just use shared_ptr?

Reply:
or references if they wont be kept around